payment/src/main/java/org/killbill/billing/payment/core/ProcessorBase.java

```diff
@@ -134,6 +136,15 @@ protected PaymentPluginApi getPaymentProviderPlugin(final UUID paymentMethodId,
         return getPaymentPluginApi(methodDao.getPluginName());
     }
 
+    protected String getPaymentProviderPluginName(final UUID paymentMethodId, final InternalTenantContext context) throws PaymentApiException {
+        final PaymentMethodModelDao methodDao = paymentDao.getPaymentMethodIncludedDeleted(paymentMethodId, context);
+        if (methodDao == null) {
+            log.error("PaymentMethod does not exist", paymentMethodId);
```

Comment:
It is not really an extra DAO call. The parts of Kill Bill that call getPaymentProviderPluginName were calling getPaymentProviderPlugin before. And this contained the same DAO call. The only difference is that I return the plugin name now and the calling method itself calls getPaymentPluginApi(pluginName) what happened before inside getPaymentProviderPlugin. The only difference is that the calling method now knows which plugin is used in its context.